### PR TITLE
standardize signature for Index reductions, implement nanmean for datetime64 dtypes

### PR DESCRIPTION
**Pull request by jbrockmendel**
Same basic goal as #23890, but focusing on nanops (a module I'm not all that familiar with) first, with the array methods later, instead of vice-versa.

I think the only user-facing change (i.e. need for whatsnew) is that `Series[datetime64].mean()` now works.

I need to see if there are any more tests from #23890 or #24024 that can be copied here.  

- [x] closes #24265
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
possibly closes #21583

**Comment by jbrockmendel** [comment]
test_nanops is a PITA in to troubleshoot.  code de-duplication at the expense of clarity may need re-thinking at some point.

**Comment by jreback** [comment]
needs. parameterization pass

File: pandas/core/base.py [hunk 1]
```diff
@@ -994,16 +1014,24 @@ def min(self):
...
+        ----------
+        axis : {None}
+            Dummy argument for consistency with Series
+        skipna : bool, default True
+
         See Also
         --------
         numpy.ndarray.argmin
         """
+        nv.validate_minmax_axis(axis)
```
Comment (jreback) [inline pandas/core/base.py]: should think about wether we can / should move these checks into the nanops themselves

File: pandas/core/dtypes/missing.py [hunk 2]
```diff
@@ -108,7 +109,7 @@ def _isna_new(obj):
     elif isinstance(obj, ABCMultiIndex):
         raise NotImplementedError("isna is not defined for MultiIndex")
     elif isinstance(obj, (ABCSeries, np.ndarray, ABCIndexClass,
-                          ABCExtensionArray)):
+                          ABCExtensionArray, ABCDatetimeArray)):
```
Comment (jreback) [inline pandas/core/dtypes/missing.py]: why is this needed? is ABCDatetimeArray NOT an extension array?

File: pandas/core/dtypes/missing.py [hunk 3]
```diff
@@ -196,6 +197,8 @@ def _isna_ndarraylike(obj):
         else:
             values = obj
         result = values.isna()
+    elif isinstance(obj, ABCDatetimeArray):
```
Comment (jreback) [inline pandas/core/dtypes/missing.py]: why is this not handled by the clause above?

File: pandas/core/indexes/datetimelike.py [hunk 4]
```diff
@@ -286,7 +286,7 @@ def min(self, axis=None, *args, **kwargs):
         except ValueError:
             return self._na_value
 
-    def argmin(self, axis=None, *args, **kwargs):
+    def argmin(self, axis=None, skipna=True, *args, **kwargs):
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: can these inherit the doc-string?

File: pandas/core/indexes/range.py [hunk 5]
```diff
@@ -297,12 +297,14 @@ def _minmax(self, meth):
 
         return self._start + self._step * no_steps
 
-    def min(self):
+    def min(self, axis=None, skipna=True):
         """The minimum value of the RangeIndex"""
```
Comment (jreback) [inline pandas/core/indexes/range.py]: same. why don't these inherit the doc-string

File: pandas/core/nanops.py [hunk 6]
```diff
@@ -212,6 +213,10 @@ def _get_values(values, skipna, fill_value=None, fill_value_typ=None,
             mask = isna(values)
 
     dtype = values.dtype
+    if is_datetime64tz_dtype(orig_values):
```
Comment (jreback) [inline pandas/core/nanops.py]: this is a very very odd place to do this as no other dtype specific things are here

File: pandas/core/nanops.py [hunk 7]
```diff
@@ -261,19 +266,24 @@ def _na_ok_dtype(dtype):
 
 def _view_if_needed(values):
     if is_datetime_or_timedelta_dtype(values):
-        return values.view(np.int64)
+        try:
```
Comment (jreback) [inline pandas/core/nanops.py]: then let's add view rather than do this

File: pandas/core/nanops.py [hunk 8]
```diff
@@ -457,7 +466,8 @@ def nanmean(values, axis=None, skipna=True, mask=None):
         values, skipna, 0, mask=mask)
     dtype_sum = dtype_max
     dtype_count = np.float64
-    if is_integer_dtype(dtype) or is_timedelta64_dtype(dtype):
+    if (is_integer_dtype(dtype) or is_timedelta64_dtype(dtype) or
```
Comment (jreback) [inline pandas/core/nanops.py]: use needs_i8_conversion

File: pandas/core/nanops.py [hunk 9]
```diff
@@ -466,14 +476,20 @@ def nanmean(values, axis=None, skipna=True, mask=None):
...
+            # suppress division by zero warnings
+            the_mean = the_sum / count
         ct_mask = count == 0
         if ct_mask.any():
             the_mean[ct_mask] = np.nan
     else:
         the_mean = the_sum / count if count > 0 else np.nan
 
-    return _wrap_results(the_mean, dtype)
+    fill_value = None
```
Comment (jreback) [inline pandas/core/nanops.py]: this is not appropriate to do here, rather in _wrap_results

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: previously discussed, this check is only for 1-d objects where `axis` is a dummy attribute

File: pandas/core/dtypes/missing.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py]: not until #24024

File: pandas/core/dtypes/missing.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py]: not until #24024

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: I'll take a look.  Might also get rid of `*args` while at it

File: pandas/core/nanops.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/nanops.py]: Semi-agree.  Most of `nanops` is numpy dtype/ndarray specific, so making datetime64tz work is klunky.  That's why I thought it made more sense to implement these directly in DTA in #23890.  But conditional on implementing this here in nanops, this check is necessary.

File: pandas/core/nanops.py (hunk 7 above)
Comment (jbrockmendel) [reply pandas/core/nanops.py]: also waiting on #24024.

File: pandas/core/nanops.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/core/nanops.py]: that includes PeriodDType, which isn't implemented yet (again, #23890 gets these all-at-once by implementing reductions directly on the arrays)

File: pandas/core/nanops.py (hunk 9 above)
Comment (jbrockmendel) [reply pandas/core/nanops.py]: OK I guess.  That changes the behavior of _wrap_results for the 8 _other_ places that call it though

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: Hmm the docstrings here and for the corresponding methods in base.IndexOpsMixin are kind of clunky.  This may merit a separate look, @datapythonista ?

File: pandas/core/nanops.py (hunk 6 above)
Comment (jreback) [reply pandas/core/nanops.py]: my point is this is in the wrong place. nanops already handles dtypes just not here.

File: pandas/core/nanops.py (hunk 7 above)
Comment (jreback) [reply pandas/core/nanops.py]: well, then this is going to have to wait for .view unless you do it here.

File: pandas/core/nanops.py (hunk 9 above)
Comment (jreback) [reply pandas/core/nanops.py]: see my above comments. this is completely changing the code flow.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (datapythonista) [reply pandas/core/indexes/datetimelike.py]: I don't know well what's the class hierarchy and whether makes sense to inherit. But we'll have to add `Parameters`, `Returns` and `Examples` section here if this is public.

File: pandas/core/nanops.py (hunk 7 above)
Comment (TomAugspurger) [reply pandas/core/nanops.py]: #24024 has one, but I don't know if it'll stick around. I'm pretty strongly +1 on keeping it though.

File: pandas/core/nanops.py (hunk 6 above)
Comment (jreback) [reply pandas/core/nanops.py]: so this should go in _view_if_needed (or maybe just remove that function and inline it here is ok

File: pandas/core/arrays/datetimelike.py [hunk 10]
```diff
@@ -483,6 +483,26 @@ def _maybe_mask_results(self, result, fill_value=iNaT, convert=None):
             result[self._isnan] = fill_value
         return result
 
+    # ------------------------------------------------------------------
+    # Additional array methods
+    # These are not part of the EA API, but we implement them because
+    # pandas currently assumes they're there.
+
+    def view(self, dtype=None):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: are we adding to EA interface? or just here (ok either way). @TomAugspurger ?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 10 above)
Comment (TomAugspurger) [reply pandas/core/arrays/datetimelike.py]: I don't think ExtensionArray.view makes sense. The result of that would (I think) depend on the memory layout of the concrete array backing the ExtesnionArray, and I think that ExtensionArray should remain agnostic to that.

**Comment by jbrockmendel** [comment]
If everyone is happy with this, I can get started on a follow-up that will trim reductions off of #24024.

File: pandas/core/base.py [hunk 11]
```diff
@@ -951,22 +957,36 @@ def max(self):
         >>> idx.max()
         ('b', 2)
         """
+        nv.validate_minmax_axis(axis)
         return nanops.nanmax(self.values)
```
Comment (TomAugspurger) [inline pandas/core/base.py]: Why isn't skipna passed? Is that a followup item?

File: pandas/core/base.py [hunk 12]
```diff
@@ -951,22 +957,36 @@ def max(self):
...
+        axis : {None}
+            Dummy argument for consistency with Series
+        skipna : bool, default True
+
         See Also
         --------
         numpy.ndarray.argmax
         """
+        nv.validate_minmax_axis(axis)
         return nanops.nanargmax(self.values)
```
Comment (TomAugspurger) [inline pandas/core/base.py]: Doesn't this hit https://github.com/pandas-dev/pandas/issues/24265?

File: pandas/core/base.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: I had considered this a follow-up item, but am now leaning towards fixing these now.  Good catch.

Comment (jbrockmendel) [reply pandas/core/base.py]: This turns out to be more involved than expected, requires some changes in Series._reduce and IndexOpsMixin._reduce.  Neither of which is all that invasive, but changing them entails changing/fixing/testing behavior for other reductions I hadn't planned on doing in this PR.

Currently going through to get that all done here.  I'm also OK with doing this in two phases so as to take things out of #24024 sooner rather than later.

File: pandas/core/base.py (hunk 12 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: Not this line, but the relevant line in indexes.datetimelike does.  With the commit just pushed, this fixed 24265

**Comment by jbrockmendel** [comment]
Updated to pass skipna in the appropriate places in Index and DatetimeIndexOpsMixin methods.  Updated a handful of tests, but I'm not confident we got thorough coverage

**Comment by jbrockmendel** [comment]
Travis fail is hypothesis

**Comment by jbrockmendel** [comment]
rebased

File: pandas/core/base.py [hunk 13]
```diff
@@ -957,10 +957,16 @@ def _ndarray_values(self):
     def empty(self):
         return not self.size
 
-    def max(self):
+    def max(self, axis=None, skipna=True):
```
Comment (jreback) [inline pandas/core/base.py]: use **kwargs I think is consistent with how we do this elsewhere, IOW we don't explicity list the numpy support args (basically remove axis)

File: pandas/core/base.py [hunk 14]
```diff
@@ -957,10 +957,16 @@ def _ndarray_values(self):
...
         """
         Return the maximum value of the Index.
 
+        Parameters
```
Comment (jreback) [inline pandas/core/base.py]: also we have similar things in pandas/core/base, these should change as well.

File: pandas/core/indexes/datetimelike.py [hunk 15]
```diff
@@ -307,23 +313,28 @@ def max(self, axis=None, *args, **kwargs):
...
+        i8 = self.asi8
         try:
-            i8 = self.asi8
-
             # quick check
             if len(i8) and self.is_monotonic:
                 if i8[-1] != iNaT:
                     return self._box_func(i8[-1])
 
+            if not len(self):
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: elif

File: pandas/core/indexes/datetimelike.py [hunk 16]
```diff
@@ -307,23 +313,28 @@ def max(self, axis=None, *args, **kwargs):
...
+                return self._na_value
+
             if self.hasnans:
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: same

File: pandas/core/nanops.py (hunk 6 above)
Comment (jreback) [reply pandas/core/nanops.py]: did you do this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reductions/test_reductions.py [hunk 17]
```diff
@@ -82,19 +113,27 @@ def test_nanops(self):
...
         assert obj.argmax() == -1
+        assert obj.argmin(skipna=False) == -1
+        assert obj.argmax(skipna=False) == -1
 
         obj = Index([pd.NaT, datetime(2011, 11, 1), datetime(2011, 11, 2),
                      pd.NaT])
         assert obj.argmin() == 1
         assert obj.argmax() == 2
+        assert obj.argmin(skipna=False) == -1
+        assert obj.argmax(skipna=False) == -1
```
Comment (jreback) [inline pandas/tests/reductions/test_reductions.py]: really? this doesn't seem right

File: pandas/tests/reductions/test_reductions.py [hunk 18]
```diff
@@ -558,6 +599,8 @@ def test_minmax_nat_series(self, nat_ser):
         # GH#23282
         assert nat_ser.min() is pd.NaT
         assert nat_ser.max() is pd.NaT
```
Comment (jreback) [inline pandas/tests/reductions/test_reductions.py]: can parameterize over skipna in a lot of the code you added to tests

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reductions/test_reductions.py (hunk 18 above)
Comment (jbrockmendel) [reply pandas/tests/reductions/test_reductions.py]: We just implemented test_reductions, still need to move over the reduction tests from tests.indexes and tests.frames.  After they're all in one place is when to really focus on parametrization.

File: pandas/tests/reductions/test_reductions.py (hunk 17 above)
Comment (jbrockmendel) [reply pandas/tests/reductions/test_reductions.py]: What would you expect it to be?  For Series idxmin/idxmax would give np.nan

File: pandas/core/base.py (hunk 13 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: Adding kwargs kinda makes sense (as long as we dont ignore them silently), but taking axis out seems weird; a big part of the goal is to make the signatures match across the different classes

File: pandas/core/nanops.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/nanops.py]: Haven't yet.  view_if_needed is called ~25 lines below this.  It isn't obvious to me that it is OK to move it up to here.  If it is, then sure, its a one-liner that can be inlined.

File: pandas/core/base.py (hunk 13 above)
Comment (jbrockmendel) [reply pandas/core/base.py]: right, the series reduction method have `axis` in their sigatures

```
>>> inspect.getargspec(pd.Series.max)
ArgSpec(args=['self', 'axis', 'skipna', 'level', 'numeric_only'], varargs=None, keywords='kwargs', defaults=(None, None, None, None))
```

File: pandas/core/nanops.py (hunk 6 above)
Comment (jreback) [reply pandas/core/nanops.py]: yes move it, as I said above you have dtype inference in 2 places for no reason.

File: pandas/core/series.py [hunk 19]
```diff
@@ -3493,6 +3494,9 @@ def _reduce(self, op, name, axis=0, skipna=True, numeric_only=None,
         # dispatch to ExtensionArray interface
         if isinstance(delegate, ExtensionArray):
             return delegate._reduce(name, skipna=skipna, **kwds)
+        elif is_datetime64_dtype(delegate):
```
Comment (jreback) [inline pandas/core/series.py]: should this be first?

File: pandas/tests/reductions/test_reductions.py [hunk 20]
```diff
@@ -56,23 +56,54 @@ def test_nanops(self):
...
                 assert getattr(obj, opname)(), datetime(2011, 11, 1)
+                assert getattr(obj, opname)(skipna=False) is pd.NaT
+
+                assert getattr(obj, arg_op)() == 1
+                result = getattr(obj, arg_op)(skipna=False)
+                if klass is Series:
+                    assert np.isnan(result)
+                else:
+                    assert result == -1
+
```
Comment (jreback) [inline pandas/tests/reductions/test_reductions.py]: put this test on the list for splitting / parametrizing later

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reductions/test_reductions.py (hunk 20 above)
Comment (jbrockmendel) [reply pandas/tests/reductions/test_reductions.py]: agreed

File: pandas/core/series.py (hunk 19 above)
Comment (jbrockmendel) [reply pandas/core/series.py]: it isn't clear to me that it would make a difference

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py [hunk 21]
```diff
@@ -957,10 +957,16 @@ def _ndarray_values(self):
...
+        ----------
+        axis : {None}
+            Dummy argument for consistency with Series
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/core/base.py]: I think @jbrockmendel is listing the set of possible values, which we do use. In this case, that set is just the single value `None`, so it looks strange.

I think

```
axis : int, optional
    For compatibility with NumPy. Only 0 or None are allowed.
```

is clearer.

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/base.py]: @TomAugspurger verbiage would be good

**Comment by jreback** [comment]
code looks good. if you can update the doc-strings then would be good.

**Comment by jbrockmendel** [comment]
I think docstring comments have been addressed; bears double-checking

**Review by jreback — APPROVED** [review]
